### PR TITLE
Multi Flow Pump v1.1

### PR DESCRIPTION
**Pull request by BrendanSanderson**
# RFC: Multi Flow v1.1.0

## Author

Brendan Sanderson

## Summary

- Modify the cap reserves functionality in Multi Flow to have two separate caps:
    1. A cap on the maximum change in exchange rate between any two tokens in the Pump
    2. A cap on the maximum change in support of LP Tokens in a Well
    
    This allows the Multi Flow Pump to independently control the maximum change in total tokens in a Well and the exchange rate between them.
- Move all constants to Pump Data to allow customizability between Wells.

## Motivation

When the BEAN:3CRV liquidity was migrated to BEAN:ETH, it took a significant time (~14 hours) for the Multi Flow capped reserves to catch up to the real balances (see [EBIP-9](https://github.com/BeanstalkFarms/Beanstalk/pull/669)). During this time, the ratio between the capped reserves did not change as they both increased by the max percent each Season. This introduced a problem where the exchange rates in Multi Flow Pump may not reflect the actual exchange rates in the pool. Multi Flow Pump needs a way to change the exchange rates even when the magnitude of the increase of reserves is capped.

Currently Multi Flow Pump's constants are stored in the Pump. However, different types of Wells (e.g., A Well with Stablecoins) may want to use different constants. Currently, this requires deploying different Pumps for each. By moving the constants to Pump Data, it will allow each Well to chose its own constants while using the same Pump.

### Given Constants:

```math 
\begin{bmatrix}
0 & \delta_{01} & \cdots & \delta_{0n} \\
\vdots & \vdots & \vdots & \vdots \\
\delta_{n0} & \delta_{n1} & \cdots & 0
\end{bmatrix}
``` 
where $\delta_{ij}$ is the maximum % increase in exchange rate of exchanging $j$ for $i$.
- $\phi_+$, $\phi_-$ be the maximum % increase and decrease in k
- $l$ be the timestamp that the Pump was last updated
- $t$ be the current timestamp
- $b$ be the block time in seconds
- $w$ be the Well Data
- $n$ be the number of tokens in the pool

### Given Functions

- `calcLpTokenSupply` - Gets the LP token supply given a list of reserves.
- `calcReserveAtRatioSwap` - Calculates the $j$th reserve such that $\prod_{i | i != j}^n \frac{d  x_j}{d x_i} = \prod_{i | i != j}^n \frac{r_j}{r_i}$ where $r$ is a list of input exchange ratios.
- `calcLPTokenUnderlying` - Calculates the amount of each reserve token underlying a given amount of LP tokens.
- `calcRate` - Calculates the exchange rate of exchange $j$ for $i$.

### Given Variables:

- $[x_{0,t}, \cdots, x_{n-1,t}]$
- $[x_{0,l}^{last}, \cdots,  x_{n-1,l}^{last}]$

### Solve For:

- $[x_{0,t}^{last}, \cdots, x_{n-1,t}^{last}]$

### Step 1: Cap Change of Magnitude

Let:

```math
k^{last}_l = \text{calcLpTokenSupply}([x_{0,l}^{last}, \cdots, x_{n-1,l}^{last}], w)
```

```math
k_t = \text{calcLpTokenSupply}([x_{0,t}, \cdots, x_{n-1,t}], w)
```

```math
k_t^{max} = k^{last}_l(1 - \phi_+)^{\frac{t-l}{b}}
```

```math
k_t^{min} = k^{last}_l(1 - \phi_-)^{\frac{t-l}{b}}
```

```math
[x_{0,t}^{cap}, \cdots, x_{n-1,t}^{cap}] = \begin{cases}
\text{calcLPTokenUnderlying}(k_t^{max}, [x_{0,t}, \cdots, x_{n-1,t}], k_t, w) & k_t > k_t^{max} \\
\text{calcLPTokenUnderlying}(k_t^{min}, [x_{0,t}, \cdots, x_{n-1,t}], k_t, w) & k_t < k_t^{min} \\
[x_{0,t}, \cdots, x_{n-1,t}] & \text{otherwise}
\end{cases}
```

### Step 2: Cap Change of Rates

For each unique set of 2 tokens in the Well (order doesn’t matter):

Let $(i, j)$ be defined such that $x_{i,t}^{cap} > x_{j,t}^{cap}$

Let:

```math
r_{ij, l}^{last} = \text{calcRate}(x_{i,l}^{last}, x_{j,l}^{last}, i, j)
```

```math
r_{ij,t}^{max} = r_{ij,l}^{last}(1 + \delta_{ij})^{\frac{t-l}{b}}
```

```math
r_{ij,t}^{min} = r_{ij,l}^{last}(1 - \frac{1}{1+\delta_{ji}})^{\frac{t-l}{b}}
```

```math
\textbf{R}_{max} = \begin{cases} 
[r^{max}_{ij,t}, 1] & i =0 \\
[1, r^{max}_{ij,t}] & otherwise
\end{cases}
```

```math
\textbf{R}_{min} = \begin{cases} 
[r^{min}_{ij,t}, 1] & i =0 \\
[1, r^{min}_{ij,t}] & otherwise
\end{cases}
```

```math
x^{\text{last}}_{i,t} = \begin{cases} \text{calcReserveAtRatioSwap}([x_{i,t}^{cap}, x_{j,t}^{cap}], i, \mathbf{R}^{max}, w)  & \frac{x_{i,t}^{cap}}{x_{j,t}^{cap}} > r_{ij,t}^{max} \\ 
\text{calcReserveAtRatioSwap}([x_{i,t}^{cap}, x_{j,t}^{cap}], i, \mathbf{R}^{min}, w) & \frac{x_{j,t}^{cap}}{x_{i,t}^{cap}} < r_{ij,t}^{min} \\
x_{i,t}^{cap} & \text{otherwise}
\\\end{cases}
```

```math
x^{\text{last}}_{j,t} = \begin{cases} \text{calcReserveAtRatioSwap}([x_{i,t}^{cap}, x_{j,t}^{cap}], j, \mathbf{R}^{max}, w)  & \frac{x_{i,t}^{cap}}{x_{j,t}^{cap}} > r_{ij,t}^{max} \\ 
\text{calcReserveAtRatioSwap}([x_{i,t}^{cap}, x_{j,t}^{cap}], j, \mathbf{R}^{min}, w) & \frac{x_{j,t}^{cap}}{x_{i,t}^{cap}} < r_{ij,t}^{min} \\
x_{j,t}^{cap} & \text{otherwise}
\\\end{cases}
```

#### 1. Create a new interface `IMultiFlowPump.sol`

```
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.20;

import {IWellFunction} from "src/interfaces/IWellFunction.sol";

/**
 * @title IMultiFlowPumpWellFunction
 * @dev A Well Function must implement IMultiFlowPumpWellFunction to be supported by
 * the Multi Flow Pump.
 */
interface IMultiFlowPumpWellFunction is IWellFunction {
    /**
     * @notice Calculates the `j` reserve such that `π_{i | i != j} (d reserves_j / d reserves_i) = π_{i | i != j}(ratios_j / ratios_i)`.
     * assumes that reserve_j is being swapped for other reserves in the Well.
     * @dev used by Beanstalk to calculate the deltaB every Season
     * @param reserves The reserves of the Well
     * @param j The index of the reserve to solve for
     * @param ratios The ratios of reserves to solve for
     * @param data Well function data provided on every call
     * @return reserve The resulting reserve at the jth index
     */
    function calcReserveAtRatioSwap(
        uint256[] calldata reserves,
        uint256 j,
        uint256[] calldata ratios,
        bytes calldata data
    ) external view returns (uint256 reserve);

    /**
     * @notice Calculates the rate at which j can be exchanged for i.
     * @param reserves The reserves of the Well
     * @param i The index of the token for which the output is being calculated
     * @param j The index of the token for which 1 token is being exchanged
     * @param data Well function data provided on every call
     * @return rate The rate at which j can be exchanged for i
     */
    function calcRate(
        uint256[] calldata reserves,
        uint256 i,
        uint256 j,
        bytes calldata data
    ) external view returns (uint256 rate);
}
```

- `IBeanstalkWellFunction` should now extend `IMultiFlowPumpWellFunction`

#### 2. Implement `IMultiFlowPump` in `ConstantProduct2`

The only function that needs to be added is `calcRate`. `calcRate` should return the output of `reserves[i] / reserves[j]`.

#### 1. Remove existing constants

- `LOG_MAX_INCREASE`
- `LOG_MAX_DECREASE`
- `ALPHA`

#### 2. Add new constants as Pump Data

- Add constants:
```
bytes16 alpha, uint256 capInterval, CapReservesParameters memory crp;

struct CapReservesParameters {
    bytes16[][] maxRatioChanges;
    bytes16 maxLpSupplyIncrease;
    bytes16 maxLpSupplyDecrease;
}
```

### CODE REVIEW

File: src/pumps/MultiFlowPump.sol [hunk 1]
```diff
@@ -169,83 +167,184 @@ contract MultiFlowPump is IPump, IMultiFlowPumpErrors, IInstantaneousPump, ICumu
...
+        if (cappedReserves.length == 0) {
+            cappedReserves = _capRatios(lastReserves, reserves, capExponent, crp, mfpWf, wf.data);
+
+            cappedReserves = _capLpTokenSupply(lastReserves, cappedReserves, capExponent, crp, mfpWf, wf.data, false);
+        } else {
+            cappedReserves = _capRatios(lastReserves, cappedReserves, capExponent, crp, mfpWf, wf.data);
         }
     }
 
+    struct CapRatiosVariables {
```
Comment (Brean0) [inline src/pumps/MultiFlowPump.sol:237]: can we move this up

**Review by Brean0 — CHANGES_REQUESTED** [review]
review

File: src/interfaces/IBeanstalkWellFunction.sol [hunk 2]
```diff
@@ -2,34 +2,18 @@
 
 pragma solidity ^0.8.20;
 
-import {IWellFunction} from "src/interfaces/IWellFunction.sol";
+import {IMultiFlowPumpWellFunction} from "src/interfaces/IMultiFlowPumpWellFunction.sol";
 
 /**
  * @title IBeanstalkWellFunction
  * @notice Defines all necessary functions for Beanstalk to support a Well Function in addition to functions defined in the primary interface.
```
[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply src/interfaces/IBeanstalkWellFunction.sol:9]: In this line:
https://github.com/BeanstalkFarms/Basin/blob/6e52d279106a7324f6537b3d55ac74a0c8b4048f/src/interfaces/IBeanstalkWellFunction.sol#L13

It says:
"Note: `calcReserveAtRatioSwap` is included in {IMultiFlowPumpWellFunction}."

This was my attempt at showcasing that `calcReserveAtRatioSwap` is defined in an interface elsewhere, but is still relevant in `IBeanstalkWellFunction`. Let me know if you think more should be added.

File: src/pumps/MultiFlowPump.sol [hunk 3]
```diff
@@ -4,12 +4,18 @@ pragma solidity ^0.8.20;
...
+import {IMultiFlowPumpWellFunction} from "src/interfaces/IMultiFlowPumpWellFunction.sol";
 import {ICumulativePump} from "src/interfaces/pumps/ICumulativePump.sol";
 import {ABDKMathQuad} from "src/libraries/ABDKMathQuad.sol";
 import {LibBytes16} from "src/libraries/LibBytes16.sol";
 import {LibLastReserveBytes} from "src/libraries/LibLastReserveBytes.sol";
+import {Math} from "oz/utils/math/Math.sol";
+import {SafeCast} from "oz/utils/math/SafeCast.sol";
+import {LibMath} from "src/libraries/LibMath.sol";
+
+// import {console} from "forge-std/console.sol";
```
[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply src/pumps/MultiFlowPump.sol]: Fixed here:
https://github.com/BeanstalkFarms/Basin/pull/127/commits/4b377359530c4675c5cdae6438dfb659d292acc4

File: src/pumps/MultiFlowPump.sol [hunk 4]
```diff
@@ -25,49 +31,42 @@ import {LibLastReserveBytes} from "src/libraries/LibLastReserveBytes.sol";
...
     using ABDKMathQuad for uint256;
+    using SafeCast for int256;
+    using Math for uint256;
+    using LibMath for uint256;
 
-    bytes16 public immutable LOG_MAX_INCREASE;
-    bytes16 public immutable LOG_MAX_DECREASE;
-    bytes16 public immutable ALPHA;
-    uint256 public immutable CAP_INTERVAL;
+    uint256 CAP_PRECISION = 1e18;
```
[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply src/pumps/MultiFlowPump.sol]: Fixed Here:
https://github.com/BeanstalkFarms/Basin/pull/127/commits/4b377359530c4675c5cdae6438dfb659d292acc4

File: src/pumps/MultiFlowPump.sol [hunk 5]
```diff
@@ -85,18 +84,21 @@ contract MultiFlowPump is IPump, IMultiFlowPumpErrors, IInstantaneousPump, ICumu
...
-            alphaN = ALPHA.powu(deltaTimestamp);
+            alphaN = alpha.powu(deltaTimestamp);
             deltaTimestampBytes = deltaTimestamp.fromUInt();
-            // Round up in case CAP_INTERVAL > block time to guarantee capExponent > 0 if time has passed since the last update.
-            capExponent = ((deltaTimestamp - 1) / CAP_INTERVAL + 1).fromUInt();
+            // Round up in case capInterval > block time to guarantee capExponent > 0 if time has passed since the last update.
+            capExponent = ((deltaTimestamp - 1) / capInterval + 1);
         }
 
+        (numberOfReserves, pumpState.lastTimestamp, pumpState.lastReserves) = slot.readLastReserves();
```
[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply src/pumps/MultiFlowPump.sol:97]: Removed here:
https://github.com/BeanstalkFarms/Basin/pull/128/commits/83d1a07fe3084eb4b17ca9d1c9a57902b24638e7

File: src/pumps/MultiFlowPump.sol [hunk 6]
```diff
@@ -169,83 +167,184 @@ contract MultiFlowPump is IPump, IMultiFlowPumpErrors, IInstantaneousPump, ICumu
...
-            }
             return cappedReserves;
         }
 
-        bytes16 capExponent = ((deltaTimestamp - 1) / CAP_INTERVAL + 1).fromUInt();
+        uint256 capExponent = ((deltaTimestamp - 1) / capInterval + 1);
+        cappedReserves = _capReserves(well, cappedReserves, currentReserves, capExponent, crp);
+    }
+
+    function _capReserves(
```
[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply src/pumps/MultiFlowPump.sol:205]: Added here:
https://github.com/BeanstalkFarms/Basin/pull/128/commits/83d1a07fe3084eb4b17ca9d1c9a57902b24638e7

File: src/pumps/MultiFlowPump.sol [hunk 7]
```diff
@@ -392,4 +502,62 @@ contract MultiFlowPump is IPump, IMultiFlowPumpErrors, IInstantaneousPump, ICumu
...
+     * If the call fails, returns the maximum possible return value for `calcLPTokenUnderlying`.
+     */
+    function tryCalcLPTokenUnderlying(
+        IMultiFlowPumpWellFunction wf,
+        uint256 lpTokenAmount,
+        uint256[] memory reserves,
+        uint256 lpTokenSupply,
+        bytes memory data
+    ) internal view returns (uint256[] memory underlyingAmounts) {
+        // return wf.calcLPTokenUnderlying(lpTokenAmount, reserves, lpTokenSupply, data);
```
[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply src/pumps/MultiFlowPump.sol:550]: Removed:
https://github.com/BeanstalkFarms/Basin/pull/128/commits/a20420ab8b72d6d1af242800c1b2cae300f2bc9f

File: src/pumps/MultiFlowPump.sol [hunk 8]
```diff
@@ -169,83 +167,184 @@ contract MultiFlowPump is IPump, IMultiFlowPumpErrors, IInstantaneousPump, ICumu
...
+            bytes16 tempExp = ABDKMathQuad.ONE.add(crp.maxRatioChanges[i][j]).powu(capExponent);
+            crv.rLimit = tempExp.cmp(MAX_CONVERT_TO_128x128) != -1
+                ? crv.rLimit = type(uint256).max
+                : crv.rLast.mulDivOrMax(tempExp.to128x128().toUint256(), CAP_PRECISION2);
+            if (cappedReserves[i] * CAP_PRECISION / cappedReserves[j] > crv.rLimit) {
+                crv.ratios = new uint256[](2);
+                crv.ratios[i] = crv.rLimit;
+                crv.ratios[j] = CAP_PRECISION;
+                // Use a minimum of 1 for reserve. Geometric means will be set to 0 if a reserve is 0.
+                // TODO: Make sure this works.
```
[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply src/pumps/MultiFlowPump.sol:272]: Done. Removed here:
https://github.com/BeanstalkFarms/Basin/pull/128/commits/3a6ed099127c827354e6381f5637c68c37d8c29b

File: src/pumps/MultiFlowPump.sol [hunk 9]
```diff
@@ -169,83 +167,184 @@ contract MultiFlowPump is IPump, IMultiFlowPumpErrors, IInstantaneousPump, ICumu
...
-                _capReserve(lastReserves[i], currentReserves[i].fromUIntToLog2(), capExponent).pow_2ToUInt();
+        Call memory wf = IWell(well).wellFunction();
+        IMultiFlowPumpWellFunction mfpWf = IMultiFlowPumpWellFunction(wf.target);
+
+        cappedReserves = _capLpTokenSupply(lastReserves, cappedReserves, capExponent, crp, mfpWf, wf.data, true);
+
+        if (cappedReserves.length == 0) {
+            cappedReserves = _capRatios(lastReserves, reserves, capExponent, crp, mfpWf, wf.data);
+
+            cappedReserves = _capLpTokenSupply(lastReserves, cappedReserves, capExponent, crp, mfpWf, wf.data, false);
```
[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply src/pumps/MultiFlowPump.sol:229]: This sequence uses a greedy procedure to maximize the precision of the output by selecting to cap the Rate or the LP token supply first.

`_capLpTokenSupply` returns an empty array if it determines that it should cap the Ratio first instead.

Line 229 will return a non-empty `cappedReserves` array after it caps the ratio and thus the length of capped Reserves will be > 0 when it calls `_capLpTokenSupply` on line 231.

[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply src/pumps/MultiFlowPump.sol:229]: Added documentation here and used `reserves ` in the call to `_capLpTokenSupply` as well:
https://github.com/BeanstalkFarms/Basin/pull/128/commits/b88d7f29f595648a4f2f04d362d4c5a92ec7cebb

File: src/pumps/MultiFlowPump.sol [hunk 10]
```diff
@@ -85,18 +84,21 @@ contract MultiFlowPump is IPump, IMultiFlowPumpErrors, IInstantaneousPump, ICumu
...
             uint256 deltaTimestamp = _getDeltaTimestamp(pumpState.lastTimestamp);
             // If no time has passed, don't update the pump reserves.
             if (deltaTimestamp == 0) return;
-            alphaN = ALPHA.powu(deltaTimestamp);
+            alphaN = alpha.powu(deltaTimestamp);
             deltaTimestampBytes = deltaTimestamp.fromUInt();
-            // Round up in case CAP_INTERVAL > block time to guarantee capExponent > 0 if time has passed since the last update.
-            capExponent = ((deltaTimestamp - 1) / CAP_INTERVAL + 1).fromUInt();
+            // Round up in case capInterval > block time to guarantee capExponent > 0 if time has passed since the last update.
+            capExponent = ((deltaTimestamp - 1) / capInterval + 1);
```
[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply src/pumps/MultiFlowPump.sol:94]: changed here:
https://github.com/BeanstalkFarms/Basin/pull/128/commits/b88d7f29f595648a4f2f04d362d4c5a92ec7cebb

File: src/pumps/MultiFlowPump.sol (hunk 1 above)
Comment (BrendanSanderson) [reply src/pumps/MultiFlowPump.sol:235]: Fixed here:
https://github.com/BeanstalkFarms/Basin/pull/128/commits/4a2f50ee54b42155085ec2f01baadb63bc41eccc

File: src/pumps/MultiFlowPump.sol [hunk 11]
```diff
@@ -169,83 +167,184 @@ contract MultiFlowPump is IPump, IMultiFlowPumpErrors, IInstantaneousPump, ICumu
...
+            bytes16 tempExp = ABDKMathQuad.ONE.add(crp.maxLpSupplyIncrease).powu(capExponent);
+            uint256 maxLpTokenSupply = tempExp.cmp(MAX_CONVERT_TO_128x128) != -1
+                ? type(uint256).max
+                : lastLpTokenSupply.mulDiv(tempExp.to128x128().toUint256(), CAP_PRECISION2);
+
+            if (lpTokenSupply > maxLpTokenSupply) {
+                // If `_capLpTokenSupply` decreases the reserves, cap the ratio first, to maximize precision.
+                if (returnIfBelowMin) return new uint256[](0);
+                cappedReserves = tryCalcLPTokenUnderlying(mfpWf, maxLpTokenSupply, cappedReserves, lpTokenSupply, data);
+                if (cappedReserves[0] == 0) cappedReserves[0] = 1;
```
[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply src/pumps/MultiFlowPump.sol:326]: Changed here:
https://github.com/BeanstalkFarms/Basin/pull/128/commits/c37e6fc1e4895ae40647fb82cca69c69ee1a2b68

File: src/pumps/MultiFlowPump.sol [hunk 12]
```diff
@@ -169,83 +167,184 @@ contract MultiFlowPump is IPump, IMultiFlowPumpErrors, IInstantaneousPump, ICumu
...
+        CapRatiosVariables memory crv;
+        crv.rLast = mfpWf.calcRate(lastReserves, i, j, data);
+
+        // If the ratio increased, check that it didn't increase above the max.
+        if (crv.r > crv.rLast) {
+            bytes16 tempExp = ABDKMathQuad.ONE.add(crp.maxRatioChanges[i][j]).powu(capExponent);
+            crv.rLimit = tempExp.cmp(MAX_CONVERT_TO_128x128) != -1
+                ? crv.rLimit = type(uint256).max
+                : crv.rLast.mulDivOrMax(tempExp.to128x128().toUint256(), CAP_PRECISION2);
+            if (cappedReserves[i] * CAP_PRECISION / cappedReserves[j] > crv.rLimit) {
```
[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply src/pumps/MultiFlowPump.sol:267]: Fixed here:
https://github.com/BeanstalkFarms/Basin/pull/128/commits/f46d5ff41b35131b2ad1cfb8c0b180e343c66aa3

File: src/pumps/MultiFlowPump.sol [hunk 13]
```diff
@@ -169,83 +167,184 @@ contract MultiFlowPump is IPump, IMultiFlowPumpErrors, IInstantaneousPump, ICumu
...
+            if (cappedReserves[i] * CAP_PRECISION / cappedReserves[j] < crv.rLimit) {
+                crv.ratios = new uint256[](2);
+                crv.ratios[i] = crv.rLimit;
+                crv.ratios[j] = CAP_PRECISION;
+                // Use a minimum of 1 for reserve. Geometric means will be set to 0 if a reserve is 0.
+                uint256 cappedReserveI =
+                    Math.max(tryCalcReserveAtRatioSwap(mfpWf, cappedReserves, i, crv.ratios, data), 1);
+                cappedReserves[j] = Math.max(tryCalcReserveAtRatioSwap(mfpWf, cappedReserves, j, crv.ratios, data), 1);
+                cappedReserves[i] = cappedReserveI;
+            }
```
[the remark this answers is not in the document]
Comment (BrendanSanderson) [reply src/pumps/MultiFlowPump.sol:294]: Wrapped here:
https://github.com/BeanstalkFarms/Basin/pull/128/commits/c13a263ec9f1c62c1fd5482808156b4f342d88df

File: src/interfaces/IBeanstalkWellFunction.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (Brean0) [reply src/interfaces/IBeanstalkWellFunction.sol:9]: okie

File: src/pumps/MultiFlowPump.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (Brean0) [reply src/pumps/MultiFlowPump.sol:229]: lgtm

**Review by Brean0 — APPROVED** [review]

**Review by Brean0 — APPROVED** [review]

**Pull request merged**